File: android/app/src/gps/java/covidsafepaths/gps/storage/RealmSecureStorage.kt
// For testing purposes you may need to retrieve data from this encrypted Relm.io database
// Directions on how to export the encryption key and use it to open the database file:
// https://pathcheck.atlassian.net/wiki/spaces/TEST/pages/50824879/Secure+Storage+Testing
// -or-
// https://gist.github.com/troach-sf/f257bb7b80e6dddd4f3bade81b7b1410


package covidsafepaths.gps.storage

import android.util.Base64
import android.util.Log
import androidx.annotation.VisibleForTesting
import com.bottlerocketstudios.vault.SharedPreferenceVault
import com.bottlerocketstudios.vault.SharedPreferenceVaultFactory
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableNativeArray
import com.marianhello.bgloc.data.BackgroundLocation
import io.realm.Realm
import io.realm.RealmConfiguration
import io.realm.Sort.ASCENDING
import io.realm.Sort.DESCENDING
import io.realm.kotlin.where
import covidsafepaths.gps.MainApplication
import covidsafepaths.gps.storage.Location.Companion.areLocationsNearby
import covidsafepaths.gps.storage.Location.Companion.createAssumedLocation
import covidsafepaths.util.getCutoffTimestamp
import java.security.SecureRandom
import java.util.UUID

open class RealmSecureStorage(inMemory: Boolean?) {
  constructor(): this(false)

  companion object {
    private const val TAG = "RealmSecureStorage"
    private const val MINIMUM_TIME_INTERVAL = 60000 * 4
    const val LOCATION_INTERVAL: Long = 60000 * 5
    private const val MAX_BACKFILL_TIME = 60000 * 60 * 24
    private const val DAYS_TO_KEEP = 14
    private const val SCHEMA_VERSION: Long = 1

    private const val MANUALLY_KEYED_PREF_FILE_NAME = "safepaths_enc_prefs"
    private const val MANUALLY_KEYED_KEY_FILE_NAME = "safepaths_enc_key"
    private const val MANUALLY_KEYED_KEY_ALIAS = "safepaths"
    private const val MANUALLY_KEYED_KEY_INDEX = 2
    private const val MANUALLY_KEYED_PRESHARED_SECRET = "" // This will not be used as we do not support < 18
    private const val KEY_REALM_ENCRYPTION_KEY = "KEY_REALM_ENCRYPTION_KEY"
  }

  private val realmConfig: RealmConfiguration

  init {
    val encryptionKey = getEncryptionKey()

    val builder = RealmConfiguration.Builder()
        .encryptionKey(encryptionKey)
        .addModule(SafePathsRealmModule())
        .schemaVersion(SCHEMA_VERSION)
        .migration(Migration())

    if (inMemory != null && inMemory) {
      builder.name(UUID.randomUUID().toString()).inMemory()
    } else {
      builder.name("safepaths.realm")
    }

    realmConfig = builder.build()
  }

  fun saveDeviceLocation(backgroundLocation: BackgroundLocation) {
    val realm = getRealmInstance()

    realm.executeTransaction {
      val realmResult = it.where<Location>()
          .equalTo(Location.KEY_SOURCE, Location.SOURCE_DEVICE)
          .sort(Location.KEY_TIME, DESCENDING)
          .limit(1)
          .findAll()

      val previousLocation = realmResult.getOrNull(0)
      val previousTime = previousLocation?.time ?: 0
      if (backgroundLocation.time - previousTime > MINIMUM_TIME_INTERVAL) {
        val assumedLocations = createAssumedLocations(previousLocation, backgroundLocation)
        if (assumedLocations.isNotEmpty()) {
          Log.d(TAG, "Inserting ${assumedLocations.size} assumed locations")
        }
        it.insert(assumedLocations)
        Log.d(TAG, "Inserting New Location")
        it.insert(Location.fromBackgroundLocation(backgroundLocation))
      } else {
        Log.d(TAG, "Ignoring save. Minimum time threshold not exceeded")
      }
    }

    realm.close()
  }

  fun importLocations(locations: ReadableArray, source: Int, promise: Promise) {
    val realm = getRealmInstance()

    val locationsToInsert = mutableListOf<Location>()
    realm.executeTransaction {
      for (i in 0 until locations.size()) {
        try {
          val map = locations.getMap(i)
          Location.fromImportLocation(map, source)?.let { location ->
            if (location.time >= getCutoffTimestamp(DAYS_TO_KEEP)) {
              locationsToInsert.add(location)
            }
          }
        } catch (exception: Exception) {
          // possible react type-safe issues here
        }
      }
      it.insertOrUpdate(locationsToInsert)
    }

    realm.close()
    promise.resolve(true)
  }

  fun trimLocations() {
    val realm = getRealmInstance()

    realm.executeTransaction {
      realm.where<Location>()
          .lessThan(Location.KEY_TIME, getCutoffTimestamp(DAYS_TO_KEEP))
          .findAll()
          .deleteAllFromRealm()
    }

    realm.close()
  }

  fun getLocations(promise: Promise) {
    val realm = getRealmInstance()

    val deviceResults = realm.where<Location>()
        .greaterThanOrEqualTo(Location.KEY_TIME, getCutoffTimestamp(DAYS_TO_KEEP))
        .sort(Location.KEY_TIME, ASCENDING)
        .findAll()

    val writeableArray = WritableNativeArray()
    deviceResults.map {
      writeableArray.pushMap(it.toWritableMap())
    }

    promise.resolve(writeableArray)
    realm.close()
  }

  fun createAssumedLocations(previousLocation: Location?, newLocation: BackgroundLocation): List<Location> {
    val assumedLocationsToInsert = mutableListOf<Location>()
    previousLocation?.let { previous ->
          val isNearbyPrevious = areLocationsNearby(previous.latitude, previous.longitude, newLocation.latitude, newLocation.longitude)
          val isTimeWithinThreshold = newLocation.time - previous.time <= MAX_BACKFILL_TIME
          if (isNearbyPrevious && isTimeWithinThreshold) {
            val newTimestamp = newLocation.time
            val latestDesiredBackfill = newTimestamp - LOCATION_INTERVAL
            val earliestDesiredBackfill = Math.max(
                newLocation.time - MAX_BACKFILL_TIME, previous.time + LOCATION_INTERVAL
            )
            for (time in latestDesiredBackfill downTo earliestDesiredBackfill step LOCATION_INTERVAL) {
              assumedLocationsToInsert.add(createAssumedLocation(time, previous.latitude, previous.longitude))
            }
          }}
    return assumedLocationsToInsert
  }

  private fun getEncryptionKey(): ByteArray {
    val vault: SharedPreferenceVault =
      SharedPreferenceVaultFactory.getAppKeyedCompatAes256Vault(
          MainApplication.getContext(), MANUALLY_KEYED_PREF_FILE_NAME,
          MANUALLY_KEYED_KEY_FILE_NAME,
          MANUALLY_KEYED_KEY_ALIAS, MANUALLY_KEYED_KEY_INDEX,
          MANUALLY_KEYED_PRESHARED_SECRET
      )

    val existingKeyString = vault.getString(KEY_REALM_ENCRYPTION_KEY, null)

    return if (existingKeyString != null) {
      Base64.decode(existingKeyString, Base64.DEFAULT)
    } else {
      val newKey = ByteArray(64)
      SecureRandom().nextBytes(newKey)
      val newKeyString = Base64.encodeToString(newKey, Base64.DEFAULT)
      vault.edit().putString(KEY_REALM_ENCRYPTION_KEY, newKeyString).apply()
      newKey
    }
  }

  @VisibleForTesting
  fun getRealmInstance(): Realm {
    return Realm.getInstance(realmConfig)
  }
}

File: ios/GPS/storage/RealmSecureStorage.swift
//
//  RealmSecureStorage.swift
//  COVIDSafePaths
//
//  Created by Tyler Roach on 5/5/20.
//  Copyright © 2020 Path Check Inc. All rights reserved.
//
// For testing purposes you may need to retrieve data from this encrypted Relm.io database
// Directions on how to export the encryption key and use it to open the database file:
// https://pathcheck.atlassian.net/wiki/spaces/TEST/pages/50824879/Secure+Storage+Testing
// -or-
// https://gist.github.com/troach-sf/f257bb7b80e6dddd4f3bade81b7b1410

import Foundation
import RealmSwift

class RealmSecureStorage {
  
  private static let DAYS_TO_KEEP = 14
  static let LOCATION_INTERVAL = 60 * 5
  private static let MINIMUM_TIME_INTERVAL = 60 * 4
  private static let MAX_BACKFILL_TIME = 60 * 60 * 24
  
  let keyLastSavedTime: String
  private let inMemory: Bool
  
  init(inMemory: Bool = false) {
    self.inMemory = inMemory
    if (inMemory) {
      self.keyLastSavedTime = "lastSavedTimeInMemory"
    } else {
      self.keyLastSavedTime = "lastSavedTime"
    }
  }
  
  func saveDeviceLocation(backgroundLocation: MAURLocation) {
    // The geolocation library sometimes returns nil times.
    // Almost immediately after these locations, we receive an identical location containing a time.
    guard backgroundLocation.time != nil else {
      return
    }
    
    // Check to only insert location if > minimum time interval
    // Using UserDefaults here since realm reads are async and we could end up saving multiple locations before a query for the most recent record returns
    let currentTime = Int(backgroundLocation.time.timeIntervalSince1970)
    let lastSavedTime = UserDefaults.standard.integer(forKey: keyLastSavedTime)
    if (currentTime - lastSavedTime < RealmSecureStorage.MINIMUM_TIME_INTERVAL) {
      return
    } else {
      UserDefaults.standard.set(currentTime, forKey: keyLastSavedTime)
    }
    
    guard let realmConfig = getRealmConfig() else { return }
    let realm = try! Realm(configuration: realmConfig)
    
    let realmResults = realm.objects(Location.self)
      .filter("\(Location.KEY_SOURCE)=\(Location.SOURCE_DEVICE)")
      .sorted(byKeyPath: Location.KEY_TIME, ascending: false)
    let previousLocation = realmResults.first
    let previousTime = previousLocation?.time ?? 0
    if currentTime - previousTime > RealmSecureStorage.MINIMUM_TIME_INTERVAL {
      let assumedLocations = createAssumedLocations(previousLocation: previousLocation, newLocation: backgroundLocation)
      try! realm.write {
        realm.add(assumedLocations, update: .modified)
      }
    }
    
    let location = Location.fromBackgroundLocation(backgroundLocation: backgroundLocation)
    try! realm.write {
      realm.add(location, update: .modified)
    }
  }
  
  func importLocations(locations: NSArray, source: Int, resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    guard let realmConfig = getRealmConfig() else {
      resolve(false)
      return
    }
    let realm = try! Realm(configuration: realmConfig)
    let locationsToInsert = locations.compactMap {
      Location.fromImportLocation(dictionary: $0 as? NSDictionary, source: source)
    }.filter { $0.time >= getCutoffTime()}
    try! realm.write {
      realm.add(locationsToInsert, update: .modified)
    }
    resolve(true)
  }
  
  func getLocations(resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    guard let realmConfig = getRealmConfig() else {
      resolve(NSArray())
      return
    }
    let realm = try! Realm(configuration: realmConfig)
    let realmResults = realm.objects(Location.self)
      .filter("\(Location.KEY_TIME)>=\(getCutoffTime())")
      .sorted(byKeyPath: Location.KEY_TIME, ascending: true)
    let shortenedLocations = Array(realmResults.map { $0.toSharableDictionary() })
    resolve(shortenedLocations)
  }
  
  func trimLocations() {
    guard let realmConfig = getRealmConfig() else {
      return
    }
    let realm = try! Realm(configuration: realmConfig)
    let realmResults = realm.objects(Location.self)
      .filter("\(Location.KEY_TIME)<\(getCutoffTime())")
    try! realm.write {
      realm.delete(realmResults)
    }
  }
  
  func createAssumedLocations(previousLocation: Location?, newLocation: MAURLocation) -> [Location] {
    var assumedLocationsToInsert: [Location] = []
    guard let previousLocation = previousLocation else {
      return assumedLocationsToInsert
    }
    
    let newLocationTime = Int(newLocation.time.timeIntervalSince1970)
    let isNearbyPrevious = Location.areLocationsNearby(
      lat1: previousLocation.latitude,
      lon1: previousLocation.longitude,
      lat2: newLocation.latitude.doubleValue,
      lon2: newLocation.longitude.doubleValue)
    let isTimeWithinThreshold = newLocationTime - previousLocation.time <= RealmSecureStorage.MAX_BACKFILL_TIME
    if isNearbyPrevious && isTimeWithinThreshold {
      let latestDesiredBackfill = newLocationTime - RealmSecureStorage.LOCATION_INTERVAL
      let earliestDesiredBackfill = max(newLocationTime - RealmSecureStorage.MAX_BACKFILL_TIME, previousLocation.time + RealmSecureStorage.LOCATION_INTERVAL)
      for time in stride(from: latestDesiredBackfill, through: earliestDesiredBackfill, by: -RealmSecureStorage.LOCATION_INTERVAL) {
        assumedLocationsToInsert.append(Location.createAssumedLocation(time: time, latitude: previousLocation.latitude, longitude: previousLocation.longitude))
      }
    }
    return assumedLocationsToInsert
  }
  
  func getCutoffTime() -> Int {
    return Int(Date().timeIntervalSince1970) - (RealmSecureStorage.DAYS_TO_KEEP * 86400)
  }
  
  func getEncyrptionKey() -> NSData? {
    // Identifier for our keychain entry - should be unique for your application
    let keychainIdentifier = "org.pathcheck.covid-safepaths.realm"
    let keychainIdentifierData = keychainIdentifier.data(using: String.Encoding.utf8, allowLossyConversion: false)!
    
    // First check in the keychain for an existing key
    var query: [NSString: AnyObject] = [
      kSecClass: kSecClassKey,
      kSecAttrApplicationTag: keychainIdentifierData as AnyObject,
      kSecAttrKeySizeInBits: 512 as AnyObject,
      kSecReturnData: true as AnyObject
    ]
    
    // To avoid Swift optimization bug, should use withUnsafeMutablePointer() function to retrieve the keychain item
    // See also: http://stackoverflow.com/questions/24145838/querying-ios-keychain-using-swift/27721328#27721328
    var dataTypeRef: AnyObject?
    var status = withUnsafeMutablePointer(to: &dataTypeRef) { SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0)) }
    if status == errSecSuccess {
      return dataTypeRef as? NSData
    }
    
    // No pre-existing key from this application, so generate a new one
    let keyData = NSMutableData(length: 64)!
    let result = SecRandomCopyBytes(kSecRandomDefault, 64, keyData.mutableBytes.bindMemory(to: UInt8.self, capacity: 64))
    assert(result == 0, "Failed to get random bytes")
    if (result != 0) {
      return nil
    }
    
    // Store the key in the keychain
    query = [
      kSecClass: kSecClassKey,
      kSecAttrApplicationTag: keychainIdentifierData as AnyObject,
      kSecAttrKeySizeInBits: 512 as AnyObject,
      kSecValueData: keyData
    ]
    
    status = SecItemAdd(query as CFDictionary, nil)
    assert(status == errSecSuccess, "Failed to insert the new key in the keychain")
    if (status != errSecSuccess) {
      return nil
    }
    return keyData
  }
  
  func getRealmConfig() -> Realm.Configuration? {
    if let key = getEncyrptionKey() {
      if (inMemory) {
        return Realm.Configuration(inMemoryIdentifier: "temp", encryptionKey: key as Data, schemaVersion: 1,
                                   migrationBlock: { _, _ in }, objectTypes: [Location.self])
      } else {
        return Realm.Configuration(encryptionKey: key as Data, schemaVersion: 1,
                                  migrationBlock: { _, _ in }, objectTypes: [Location.self])
      }
    } else {
      return nil
    }
  }
  
  func getRealmInstance() -> Realm? {
    guard let config = getRealmConfig() else {
      return nil
    }
    
    return try! Realm(configuration: config)
  }
}
